### PR TITLE
[RDY] Fix inccomand performance degradation

### PR DESCRIPTION
**Pull request by KillTheMule**
It was broken since the introduction of the macro.

Mentioned on https://www.reddit.com/r/neovim/comments/j9djia/making_inccommand_work_on_displayed_text/

Note: I added `if (preview)` around the counting code, because if we're not previewing we're not really setting the numbers right and are looking at integer underflow, which is undefined behavior afaik.

(e) Not sure if this is common knowledge, but if you change the diff view to ignore whitespace changes, it clearly shows what this is about. If you show whitespace changes, not so much...

### CODE REVIEW

**Comment by medwatt** [comment]
Is it possible to know when this fix will be merged ?

**Comment by janlazo** [comment]
@KillTheMule Can you add a test?

**Comment by KillTheMule** [comment]
> @KillTheMule Can you add a test?

Ha! Maybe? Not sure, we [had a test](https://github.com/neovim/neovim/pull/6949), that somehow didn't fail at any time, although the whole thing was broken. 

Generally speaking, a performance improvement such as this is kinda hard to test, since the timing vary so much between different systems. We don't want a test to pass on slow CI but fail on local hardware, right?

So unless we muck with the code specifically for tests (I could introduce some sort of counter, ofc), I don't really see how to test this reliably, in the light of the test above not really doing what it should... My suggestion would be to just accept that this is not a functional change and we don't really have a good way of putting up a testable benchmark here. But I'm definitely open to suggestions and would put in the work necessary to make this fly :)

**Comment by janlazo** [comment]
> Generally speaking, a performance improvement such as this is kinda hard to test, since the timing vary so much between different systems. We don't want a test to pass on slow CI but fail on local hardware, right?

It's fine if the tests run in some environments only. Some functional tests use `load_adjust`, like the test in https://github.com/neovim/neovim/pull/12665/files#diff-615ef1058bd46f3cd6d4aa9832115ac8bc52fdf307a7d83bd0e570b104d60b25, to account for slow CI.

**Comment by janlazo** [comment]
https://travis-ci.org/github/neovim/neovim/jobs/737146410

```
ninja  -C build lualint
ninja: Entering directory `build'
[1/1] cd /home/travis/build/neovim/neovim && /home/travis/nvim-deps/usr/bin/luacheck -q runtime/ src/ test/
FAILED: CMakeFiles/lualint 
cd /home/travis/build/neovim/neovim && /home/travis/nvim-deps/usr/bin/luacheck -q runtime/ src/ test/
Checking test/functional/ui/inccommand_spec.lua   2 warnings
    test/functional/ui/inccommand_spec.lua:1312:11: unused variable time_icm
    test/functional/ui/inccommand_spec.lua:1325:44: trailing whitespace in a comment
Total: 2 warnings / 0 errors in 425 files
ninja: build stopped: subcommand failed.
Makefile:148: recipe for target 'lualint' failed
make: *** [lualint] Error 1
+fail lualint
+local test_name=lualint
+local fail_char=
+local message=
+: F
+: Test lualint failed
+local 'full_msg=F /lualint|lualint :: Test lualint failed'
+FAIL_SUMMARY='
F /lualint|lualint :: Test lualint failed'
+echo 'F /lualint|lualint :: Test lualint failed'
+echo 'Failed: F /lualint|lualint :: Test lualint failed'
Failed: F /lualint|lualint :: Test lualint failed
+FAILED=1
+exit_suite --continue
+set +x
Suite /lualint failed, summary:
```

**Comment by KillTheMule** [comment]
That's very unsurprising, seeing I commented out almost everything of the test. It's just there because I'm waiting for @bfredl (or anyone else in the know) to tell me how to do [this](https://github.com/neovim/neovim/pull/13089/files#diff-9d6a7b9a4549ef72c7a3c2793f37f49649dfac113a736b72b9240d5049073103R1307) measurement properly, because `feed` does not wait for the key presses to be processed. But that's what I need to measure...

Or any other idea how to test this thing?

File: test/functional/ui/inccommand_spec.lua [hunk 1]
```diff
@@ -1279,6 +1281,52 @@ describe(":substitute, inccommand=split", function()
...
+    local before_subst = reltime()
+    for i = 1,9,1 do
+      command("%s/de"..tostring(i-1).."/de"..tostring(i).."/g")
+    end
+    local time_subst = reltimefloat(reltime(before_subst))
+
+    -- the substitution churns through 10k lines, while the preview only looks
+    -- at 12 (default of previewhight), so the following should hold on any
+    -- system, even with a lot of variation 
+    eq(time_icm, time_subst)
```
Comment (janlazo) [inline test/functional/ui/inccommand_spec.lua:1326]: Isn't this flaky even in a local environment?

Comment (KillTheMule) [reply test/functional/ui/inccommand_spec.lua:1326]: It's just for debugging stuff, I wanted to check for `time_icm * 10 < time_subst`, but `feed` isn't doing what I need (see below).

**Comment by medwatt** [comment]
Is this ever going to be resolved?

**Comment by clason** [comment]
If people only ask for an ETA without Doing The Work, no.

**Comment by medwatt** [comment]
> If people only ask for an ETA without Doing The Work, no.

Alright.

**Comment by KillTheMule** [comment]
> If people only ask for an ETA without Doing The Work, no.

On the other hand, this is a simple performance improvement. Correctness tests are still in place, so this is basically good to go. Since I don't have any idea how to properly test performance on CI here, and nobody has come up with a way, I'd simply suggest merging this without testing performance. Or are there concerns regarding correctness?

**Comment by clason** [comment]
@KillTheMule If you think this is good to go, rebasing on master would go a long way; then people could test it.

@medwatt Possible work would be "I tested this and it works fine with much better performance for me on <platform>". A comment like this would add useful information and help move this along, in a way that "fix when?" doesn't.

**Comment by KillTheMule** [comment]
Sure, I'll send a rebase later on.

**Comment by clason** [comment]
Also, could we replace the huge fixture in the test by just programmatically creating in a setup step? (Unless this was a actually a sneaky ploy for improving your commit stats ;))

**Comment by bfredl** [comment]
I'd suggest remove the test for now and let's just merge the change. Our test suite is not really tuned for performance tests at all (at least not with CI setup with wildly changing CPU load outside of our control)

**Pull request closed without merging**